index.html

```diff
@@ -86,25 +86,311 @@ <h3>Use-case #4: Avoiding "Hall of Mirrors"</h3>
+        </p>
+        <pre class="idl">
+          partial interface MediaStreamTrack {
+            sequence&lt;DOMString&gt; getSupportedCaptureActions();
+            Promise&lt;undefined&gt; sendCaptureAction(CaptureAction action);
+          };
+        </pre>
+        <dl data-link-for="MediaStreamTrack" data-dfn-for="MediaStreamTrack">
+          <dt>
+            <dfn>getSupportedCaptureActions</dfn>
+          </dt>
+          <dd>
+            <p>When this method is invoked, the user agent MUST return [=this=]'
+            {{MediaDevices/[[AvailableCaptureActions]]}} if defined, or `[]` if not defined.</p>
+          </dd>
+          <dt>
+            <dfn>sendCaptureAction</dfn>
+          </dt>
+          <dd>
+            <p>When this method is invoked, the user agent MUST run the following steps:</p>
+            <ol>
+              <li>
+                If the [=relevant global object=] of [=this=] does not have
+                [=transient activation=], return a promise [=rejected=] with
+                {{InvalidStateError}}.
+              </li>
+              <li>
+                [=Consume user activation=].
+              </li>
+              <li>
+                Let |action| be the method's first argument.
+              </li>
+              <li>
+                If |action| is not in [=this=]' {{MediaDevices/[[AvailableCaptureActions]]}},
+                return a promise [=rejected=] with {{NotFoundError}}.
+              </li>
+              <li>
+                Let |p| be a new promise.
+              </li>
+              <li>
+                Run the following steps in parallel:
+                <ol>
+                  <li>
+                    <p>
+                      Queue a task on the task-list of the captured
+                      <a data-cite="SCREEN-CAPTURE#dfn-browser">browser</a>
+                      <a data-cite="SCREEN-CAPTURE#dfn-display-surface">display surface</a>'s
+                      [=top-level browsing context=]'s [=active document=] to run the
+                      following steps:
+                    </p>
+                    <ol>
+                      <li>
+                        Let |target| be the the [=relevant settings object=]'s
+                        [=environment settings object/responsible document=]'s
+                        associated navigator's {{MediaDevices}} object.
+                      <li>
+                        If |action| is not in |target|'s
+                        {{MediaDevices/[[RegisteredCaptureActions]]}}, abort these steps.
+                      <li>
+                        [=Fire an event=] named `"captureaction"`, using a
+                        {{CaptureActionEvent}} with {{CaptureActionEventInit/action}}
+                        set to |action|, at |target|.
+                      </li>
+                    </ol>
+                    <li>
+                      Wait for the event to have been fired.
```

Comment:
So the capturer gets a glimpse into the event loop of the captured application? Sounds iffy. Do we have a compelling reason motivating this? Seems to me like a user flipping slides on a captured deck can just visually observe when the slide changes. Why does the application need a Promise, etc.?

Reply:
I don't see how this looks into anyone's event loop. This lets the capturer know that the action has been delivered.

At minimum, I'd expect capturers would not want to send another action until then. I could also see it wanting to e.g. show a (best-guess) slide number or progress bar or whatever. But if it updates out of lockstep with when the actual slide flips, it might look janky. A promise might not solve this 100% but gives them something to work with.

This spec also enables out-of-band communication between capturer and capturee, so this ACK might help avoid races.

Reply:
> This lets the capturer know that the action has been delivered.

Your current design only allows declared actions to be accepted. Since this set of declared actions can change at any moment (explicitly or by navigation), validation of delivery requires synchronizing with the receiver's event loop.

> I could also see it wanting to e.g. show a (best-guess)

I find it hard to believe any VC software is going to show a guess of the slide number.